File: Algorithms/src/3100. Water Bottles II/Code/solution.java
class Solution {
    public int maxBottlesDrunk(int numBottles, int numExchange) {
        int full = numBottles;    // current full bottles
        int empty = 0;            // current empty bottles
        int ans = 0;              // total bottles drunk
        int curEx = numExchange;  // current exchange threshold

        while (full > 0) {
            // drink all full bottles
            ans += full;
            empty += full;
            full = 0;

            // perform one-by-one exchanges while possible
            while (empty >= curEx) {
                empty -= curEx;  // use curEx empty bottles
                full += 1;       // receive one full bottle
                curEx += 1;      // increase exchange requirement
            }
        }
        return ans;
    }
}

File: Algorithms/src/3100. Water Bottles II/Code/solution.go
func maxBottlesDrunk(numBottles int, numExchange int) int {
    full := numBottles
    empty := 0
    ans := 0
    curEx := numExchange

    for full > 0 {
        // drink all full bottles
        ans += full
        empty += full
        full = 0

        // exchange one-by-one while possible
        for empty >= curEx {
            empty -= curEx
            full += 1
            curEx += 1
        }
    }
    return ans
}
